### PR TITLE
fix: improve physical execution for coalesce

### LINKED ISSUE #1988: Follow-up tasks for coalesce and repartition
https://github.com/lakehq/sail/pull/1856#pullrequestreview-4362485411

### PR DESCRIPTION
**Pull request by twsl**
Fix #1988

- Added `coalesce` module to the physical plan.
- Enhanced `ExplicitRepartitionExec` to support round-robin and coalesce repartitioning.
- Updated `PlanResolver` to handle repartition hints and validate partition counts.
- Introduced `RowRoundRobinPartitioner` for efficient round-robin partitioning.
- Modified tests to cover new repartitioning behaviors and validate partition counts.
- Ensured that repartitioning maintains input partition integrity and handles edge cases.

### CODE REVIEW

**Comment by twsl** [comment]
hey @linhr feel free to check this out and take over to finish this pr

**Comment by linhr** [comment]
Great! Let me take a look. I may merge it as is and I'll take care of my own comments (if any) in a follow-up PR. Thanks for having this!

**Comment by linhr** [comment]
I've triggered a Copilot review but no need to make further changes if it generates comments. I'm just using it as a reference for my review.

File: crates/sail-plan/src/resolver/query/misc.rs [hunk 1]
```diff
@@ -224,3 +222,37 @@ impl PlanResolver<'_> {
...
+        Expr::Literal(ScalarValue::UInt8(Some(value)), _metadata) => Ok(*value as usize),
+        Expr::Literal(ScalarValue::UInt16(Some(value)), _metadata) => Ok(*value as usize),
+        Expr::Literal(ScalarValue::UInt32(Some(value)), _metadata) => Ok(*value as usize),
+        Expr::Literal(ScalarValue::UInt64(Some(value)), _metadata) => usize::try_from(*value)
+            .map_err(|_| {
+                PlanError::invalid(format!("{hint_name} hint partition count is too large"))
+            }),
+        _ => Err(PlanError::invalid(format!(
+            "{hint_name} hint partition count must be an integer"
+        ))),
```
[the remark this answers is not in the document]
Comment (linhr) [reply crates/sail-plan/src/resolver/query/misc.rs:256]: Yeah I'm not sure why we switched to literal matching instead of using `LiteralEvaluator`. The latter would be useful if we want to allow literal expressions such as `1 + 1`.

File: crates/sail-physical-plan/src/repartition.rs [hunk 2]
```diff
@@ -83,10 +84,48 @@ impl RowRoundRobinPartitioner {
...
+        if state.active_streams == 0 {
+            if let Some(receivers) = state.receivers.as_mut() {
+                for receiver in receivers.iter_mut() {
+                    receiver.take();
+                }
+            }
+            state.tasks.take();
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (linhr) [reply crates/sail-physical-plan/src/repartition.rs:119]: Yeah this comment likely would need some attention.

**Review by linhr — APPROVED** [review]
This looks good to me overall! Let me merge this and then I'll play with the code a bit more and look into Copilot review comments.

File: crates/sail-physical-plan/src/repartition.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (linhr) [reply crates/sail-physical-plan/src/repartition.rs:119]: #2037
